### ISSUE TITLE
[idea] update world matrix only if local matrix composition yields a new result.

### ISSUE THREAD

**Issue by trusktr** [mrdoob/three.js#15443]
##### Description of the problem

At the moment, the `Object3D.updateMatrix` method,

https://github.com/mrdoob/three.js/blob/f55b3d3b91b0f50c28523ed707baec91ddff465a/src/core/Object3D.js#L592-L598

always causes `this.matrixWorldNeedsUpdate` to be true.

Instead it can set that to true only if the resulting composition results in a changed local matrix.

Would detecting a change be cheaper than just always multiplying the parent world matrix? My gut feeling is that it would be.

**Comment by Mugen87** [comment]
Related #14138

Different approaches were discussed in this PR, also the possibility to cache tranformation matrices similar to other engines like BabylonJS.

**Comment by mrdoob** [comment]
> Would detecting a change be cheaper than just always multiplying the parent world matrix? My gut feeling is that it would be.

Maybe you can implement it on your end and report performance numbers?

**Comment by takahirox** [comment]
FYI, I reported the performance gain/loss of benchmark test with dirty flag and cache ideas in #14138.

**Comment by mrdoob** [comment]
Yeah, I guess this is a duplicate of #14138.

**Comment by donhatch** [comment]
Hi folks.
It  looks to me like this is *not* a duplicate of #14138; that one is far more intrusive and ambitious, to the point where it turned out to be infeasible, if I am reading that issue thread correctly.

In contrast, this one has very small scope and essentially no downside, AFAICT.  This one looks, to me, like a rare piece of low-hanging non-problematic fruit that could be plucked, leaving this generally rather bleak landscape a little better.

To be clear, if I understand correctly, this proposed change is, simply, to change Object3D.updateMatrix() from this:
```
    updateMatrix() {
            this.matrix.compose( this.position, this.quaternion, this.scale );
            this.matrixWorldNeedsUpdate = true;
    }
```
to this:
```
    updateMatrix() {
        _m1.compose( this.position, this.quaternion, this.scale );
        if ( ! _m1.equals( this.matrix ) ) {
            this.matrix.copy( _m1 );
            this.matrixWorldNeedsUpdate = true;
        }
    }
```

Some details about the expected impact during WebGLRenderer.render()'s initial scene.updateMatrixWorld() top-down traversal (which is where I've observed this function and its consequences being especially hot, even when no transforms are changing at all!):
- I don't expect this will have any performance gain in most scenes, because of other broken windows elsewhere. (In fact it will have a tiny performance *loss* in most scenes, but I hope everyone will agree that this isn't significant enough to be worth measuring or proving anything about).
- In more detail, the reason is that, if I'm guessing correctly, most scenes have .matrixAutoUpdate and .matrixWorldAutoUpdate turned on throughout, since that's the default, and, when that's the case, renderer.render()'s call to scene.updateMatrixWorld() will still always do a full updateMatrixWorld() traversal of every node in the scene tree, even when no transforms changed.  The only difference will be that this full traversal will be with force=false instead of false=true, but otherwise it's essentially the same traversal.

So the real benefit of this proposed change is that it *allows* the app / scene creator to avoid the full updateMatrixWorld() traversal of the whole scene when nothing changed (or. more generally, in subtrees where nothing changed), by carefully choosing to turn off .matrixWorldAutoUpdate on some or all nodes (which necessitates the app doing its own [push-dirty / pull-clean management](https://github.com/mrdoob/three.js/pull/14138#issuecomment-392222459), probably involving setting .matrixWorldNeedsUpdate as appropriate throughout the scene, but that's another chapter in the story).

So here's the 2x2 matrix of situations (assuming .matrixAutoUpdate is true throughout):

| App pays attention to .matrixWorldAutoUpdate? | without proposed change | with proposed change |
| ----------- | ----------- | - |
| no      | full updateMatrixWorld(force=true) traversal every render  | full updateMatrixWorld(force=false) traversal every render |
| yes   | full updateMatrixWorld(force=true) traversal every render | traverses only what changed!  |

If my analysis is correct, could this be reopened and considered?

**Comment by mrdoob** [comment]
TL;DR? 😅

**Comment by donhatch** [comment]
Not a dup.  Please reopen and change the code as suggested, there's no downside!
